Move top level parsing of sort element to SortBuilder

cbuescher opened the pull request:
This moves the current top level parsing code for the whole "sort" element from SortParseElement into a static "fromXContent" method in SortBuilder. Also adding tests for parsing lists of SortBuilders in the different allowed syntax variations (e.g. { "sort" : "field1" }, { "sort" : { "field1" : "asc" }}, { "sort" : [ "field1", "field2" ] }.

cbuescher committed:
Add parsing of list of sort builders to SortBuilder

Moving the current parsing code for the whole "sort" element
in the seach source over to static "fromXContent" method in
SortBuilder.

cbuescher commented:
@MaineC I started breaking up #17205 into parts that should be better to review. This one adds only the parsing and tests for SortBuilder#fromXContent().

MaineC commented on core/src/main/java/org/elasticsearch/search/sort/SortBuilder.java:
```diff
@@ -27,47 +29,44 @@
-        }
-        return nested;
-    }
+public abstract class SortBuilder<T extends SortBuilder<?>> implements SortBuilderParser<T>, ToXContent {
```
In a followup PR we should merge SortBuilder and SortBuilderParser, I think. The latter one was only introduced as an intermediate step to avoid having to refactor all builders at once. Not sure if we can add the interface ToXContent there as well then.

cbuescher replied:
We can do it in this PR already, I think. Would you rather put fromXContent and build() in as abstract methods or do you think we should have a SortBuilder interface and make this class AbstractSortBuilder instead?

MaineC replied:
In other cases like this we went for reducing the number of classes, so here too, I'd go for adding these two simply as abstract methods.

cbuescher committed:
Remove SortBuilderParser interface, using abstract methods in SortBuilder

MaineC commented on core/src/main/java/org/elasticsearch/search/sort/SortBuilder.java:
```diff
@@ -96,4 +95,122 @@ public T order(SortOrder order) {
+                    if (fieldName.equals(ScoreSortBuilder.NAME)) {
+                        sortFields.add(new ScoreSortBuilder().order(order));
+                    } else {
+                        sortFields.add(new FieldSortBuilder(fieldName).order(order));
```
After reading this distinction for the third time - maybe generating a new FieldSortBuilder or ScoreSortBuilder could be factored into a private helper method like so:

private SortBuilder generate(String fieldName) {
...
}

cbuescher replied:
+1

MaineC commented on core/src/main/java/org/elasticsearch/search/sort/SortBuilder.java:
```diff
@@ -96,4 +95,122 @@ public T order(SortOrder order) {
+        }
+    }
+
+    public static void parseSort(XContentParser parser, SearchContext context) throws IOException {
```
Do I read this correctly that this was introduced for testing only? Don't find it anywhere else. Could we reduce visibility if so?

cbuescher replied:
Yes, this will be used to add compatibility with other users of current SortParseElement. I can take it out of this PR though.

MaineC commented on core/src/test/java/org/elasticsearch/search/sort/FieldSortBuilderTests.java:
```diff
@@ -25,10 +25,14 @@
         FieldSortBuilder builder = new FieldSortBuilder(fieldName);
         if (randomBoolean()) {
-            builder.order(RandomSortDataGenerator.order(builder.order()));
+            builder.order(RandomSortDataGenerator.order(null));
```
So - by default, order is initialised to ASC.

Before this change, calling this method would always set the order DESC, so as we are calling it randomly, half of all tests run with the original ASC, half run with DESC.

After this change, calling this method will always set the order to ASC, so all tests should be running with ASC sorting, leaving DESC untested.

So - why this change?

If this was unintentional, most likely "RandomSortDataGenerator" is a bad name for a class that returns a random value except the one provided.

MaineC replied:
After reading further - scratch that comment.

MaineC commented on core/src/test/java/org/elasticsearch/search/sort/GeoDistanceSortBuilderTests.java:
```diff
@@ -70,7 +74,7 @@ protected GeoDistanceSortBuilder createTestItem() {
         }
         if (randomBoolean()) {
-            result.order(RandomSortDataGenerator.order(result.order()));
+            result.order(RandomSortDataGenerator.order(null));
```
See above.

MaineC replied:
See answer comment above for another answer here.

MaineC commented:
Left two minor comments - up to your judgement whether to make these changes now, later or not at all.

Other than that LGTM.

cbuescher commented:
@MaineC thanks, I removed the SortBuilderParser interface and moved those two abstract methods into SortBuilder. Also adding the helper method you suggested and removing unused code that I will only need in the follow-up PR.
